models/domains_domains.php

```diff
@@ -172,4 +209,62 @@ public function updateNameservers($service_id, array $nameservers)
 
         return $result;
     }
+
+    /**
+     * Gets the domain expiration date
+     *
+     * @param int $service_id The id of the service where the domain belongs
+     * @param string $format The format to return the expiration date in
+     * @return string The domain expiration date in UTC time in the given format
+     * @see Services::get()
+     */
+    public function getExpirationDate($service_id, $format = 'Y-m-d H:i:s')
+    {
+        Loader::loadModels($this, ['Services', 'ModuleManager']);
+
+        if (is_numeric($service_id)) {
+            $service = $this->Services->get($service_id);
+
+            // Check if the domain already has an expiration date
+            $domain = $this->Record->select()
+                ->from('domains_domains')
+                ->where('service_id', '=', $service->id)
+                ->fetch();
+            $expiration_date = $domain->expiration_date ?? null;
+
+            // Get the expiration date from the registrar
+            if (empty($expiration_date)) {
+                $expiration_date = $this->ModuleManager->moduleRpc(
+                    $service->package->module_id,
+                    'getExpirationDate',
+                    [$service_id, $format],
+                    $service->module_row_id ?? null
+                );
+            }
+
+            if (empty($expiration_date)) {
+                $expiration_date = $service->date_renews;
+            }
+
+            if ($expiration_date === $service->date_renews) {
```

Comment:
I'm not a fan of incidental saves within get functions.  Let's create a separate set function:
```
    public function setExpirationDate($service_id, $expiration_date)
    {
        $this->Record->duplicate('domains_domains.expiration_date', '=', $expiration_date)
            ->insert('domains_domains', ['service_id' => $service_id, 'expiration_date' => $expiration_date]);
    }
```
with doc comments.  Then in locations where this is called but the expiration may not be set (like DomainPlugins::setRenewalDate()) we can call 
```
$expiration_date = $this->DomainsDomains->getExpirationDate($service->id);
$this->DomainsDomains->setExpirationDate($expiration_date);
```